porcelain/miner.go

```diff
@@ -12,13 +12,146 @@ import (
+	MessageWait(ctx context.Context, msgCid cid.Cid, cb func(*types.Block, *types.SignedMessage, *types.MessageReceipt) error) error
+	NetworkGetPeerID() peer.ID
+	WalletFind(address address.Address) (w.Backend, error)
+	WalletGetPubKeyForAddress(addr address.Address) ([]byte, error)
+	WalletHasAddress(addr address.Address) bool
+}
+
+// MinerCreate creates a new miner actor for the given account and returns its address.
+// It will wait for the the actor to appear on-chain and add set the address to mining.minerAddress in the config.
+// TODO: add ability to pass in a KeyInfo to store for signing blocks.
+//       See https://github.com/filecoin-project/go-filecoin/issues/1843
+func MinerCreate(
+	ctx context.Context,
+	plumbing mcAPI,
+	accountAddr address.Address,
+	gasPrice types.AttoFIL,
+	gasLimit types.GasUnits,
+	pledge uint64,
+	pid peer.ID,
+	collateral *types.AttoFIL,
+) (_ *address.Address, err error) {
+	if accountAddr == (address.Address{}) {
+		accountAddr, err = plumbing.GetAndMaybeSetDefaultSenderAddress()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if pid == "" {
+		pid = plumbing.NetworkGetPeerID()
+	}
+
+	ctx = log.Start(ctx, "Node.CreateMiner")
+	defer func() {
+		log.FinishWithErr(ctx, err)
+	}()
+
+	addrInterface, err := plumbing.ConfigGet("mining.minerAddress")
+	if err != nil {
+		return nil, err
+	}
+	addr, success := addrInterface.(address.Address)
+	if !success {
+		return nil, fmt.Errorf("failed to read miner address")
+	}
+	if (addr != address.Address{}) {
+		return nil, fmt.Errorf("can only have one miner per node")
+	}
+
+	pubKey := []byte{}
+	if plumbing.WalletHasAddress(addr) {
+		pubKey, err = plumbing.WalletGetPubKeyForAddress(addr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	smsgCid, err := plumbing.MessageSend(
+		ctx,
+		accountAddr,
+		address.StorageMarketAddress,
+		collateral,
+		gasPrice,
+		gasLimit,
+		"createMiner",
+		big.NewInt(int64(pledge)),
+		pubKey,
+		pid,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	minerAddr, err := waitForMinerAddress(ctx, plumbing, smsgCid)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = persistMinerAddress(ctx, plumbing, minerAddr); err != nil {
+		return nil, err
+	}
+
+	return &minerAddr, nil
+}
+
+func waitForMinerAddress(ctx context.Context, plumbing mcAPI, smsgCid cid.Cid) (address.Address, error) {
+	var minerAddr address.Address
+	err := plumbing.MessageWait(ctx, smsgCid, func(blk *types.Block, smsg *types.SignedMessage, receipt *types.MessageReceipt) (err error) {
+		if receipt.ExitCode != uint8(0) {
+			return vmErrors.VMExitCodeToError(receipt.ExitCode, storagemarket.Errors)
+		}
+		minerAddr, err = address.NewFromBytes(receipt.Return[0])
+		return err
+	})
+	return minerAddr, err
+}
+
+func persistMinerAddress(ctx context.Context, plumbing mcAPI, minerAddr address.Address) error {
+	// TODO: https://github.com/filecoin-project/go-filecoin/issues/1843
```

Comment:
this todo is referenced at the entrypoint above, suggest that is sufficient so remove this reminder